sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -197,10 +197,13 @@ case class DataSource(
         SparkHadoopUtil.get.globPathIfNecessary(qualified)
       }.toArray
       val fileCatalog = new ListingFileCatalog(sparkSession, globbedPaths, options, None)
+      val partitionCols = fileCatalog.partitionSpec().partitionColumns.fields
       format.inferSchema(
         sparkSession,
         caseInsensitiveOptions,
-        fileCatalog.allFiles())
+        fileCatalog.allFiles()).map { inferredSchema =>
```

Comment:
The wrapping threw me off for a minute. You might consider introducing an extra variable for the result of inferSchema. Also does `.map(partitionCols.foldLeft(_)((struct, field) => struct.add(field)))` work?